### ISSUE TITLE
Non deterministic build result for structure.

### ISSUE THREAD

**Issue by mfilippov** [dotnet/roslyn#53865]
I found inconsistent compiler behaviour when the code has a nested structure and use a field from a method. 
Code example:
```
public unsafe struct UnsafeStruct0
    {
        public fixed ushort FixedField1[10];
        public fixed ushort FixedField2[10];

        public UnsafeStruct4* pStruct;

        public void AccessMethod()
        {
            pStruct->FixedField2[0] = 0;
        }
    }
```
You need many structures like this and parallel compilation to easily reproduce the problem.
I attached a sample project to the issue.

**Version Used**:  3.9.0

**Steps to Reproduce**:

1. Run attached repro project
2. You will see inconsistent hashes
3.  Change WithConcurrentBuild parameter in Program.cs line 45 to false
4. Run again and you will see consistent hashes 

**Expected Behavior**:
Result hashes should be constant

**Actual Behavior**:
Every build produce a new hash of the assembly

This issue blocks me from creating a reproducible build.

[RoslynNondeterministicRepro.zip](https://github.com/dotnet/roslyn/files/6594240/RoslynNondeterministicRepro.zip)

[jcouv: note to self, investigating in `fixed-bug` branch]

**Comment by jcouv** [comment]
Fix is being reviewed for VS2022.

**Comment by jcouv** [comment]
Fixed in 17.0p5
